File: leetcode/0381-insert-delete-getrandom-o1-duplicates-allowed/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

class RandomizedCollection {
    private final List<Integer> values = new ArrayList<>();
    private final Map<Integer, Set<Integer>> indexes = new HashMap<>();
    private final Random rand = new Random();

    public RandomizedCollection() {
    }

    public boolean insert(int val) {
        indexes.putIfAbsent(val, new HashSet<>());
        boolean isNew = indexes.get(val).isEmpty();
        indexes.get(val).add(values.size());
        values.add(val);
        return isNew;
    }

    public boolean remove(int val) {
        Set<Integer> idxSet = indexes.get(val);
        if (idxSet == null || idxSet.isEmpty()) return false;

        int removeIdx = idxSet.iterator().next();
        idxSet.remove(removeIdx); // this occurrence of val is gone

        int lastIdx = values.size() - 1;
        int lastVal = values.get(lastIdx);

        if (removeIdx != lastIdx) {
            values.set(removeIdx, lastVal);
            // the value that used to live at lastIdx now lives at removeIdx
            indexes.get(lastVal).remove(lastIdx);
            indexes.get(lastVal).add(removeIdx);
        }

        values.remove(values.size() - 1);
        return true;
    }

    public int getRandom() {
        return values.get(rand.nextInt(values.size()));
    }
}

/**
 * Your RandomizedCollection object will be instantiated and called as such:
 * RandomizedCollection obj = new RandomizedCollection();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: leetcode/0381-insert-delete-getrandom-o1-duplicates-allowed/solution.py
import random
from collections import defaultdict


class RandomizedCollection:
    def __init__(self):
        self.values = []
        self.indexes = defaultdict(set)

    def insert(self, val: int) -> bool:
        is_new = len(self.indexes[val]) == 0
        self.indexes[val].add(len(self.values))
        self.values.append(val)
        return is_new

    def remove(self, val: int) -> bool:
        if not self.indexes[val]:
            return False

        remove_idx = next(iter(self.indexes[val]))
        self.indexes[val].discard(remove_idx)  # this occurrence of val is gone

        last_idx = len(self.values) - 1
        last_val = self.values[last_idx]

        if remove_idx != last_idx:
            self.values[remove_idx] = last_val
            # the value that used to live at last_idx now lives at remove_idx
            self.indexes[last_val].discard(last_idx)
            self.indexes[last_val].add(remove_idx)

        self.values.pop()
        return True

    def getRandom(self) -> int:
        return random.choice(self.values)


# Your RandomizedCollection object will be instantiated and called as such:
# obj = RandomizedCollection()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
